docs/rfc/rfc-003-performance-questions.md

```diff
@@ -0,0 +1,193 @@
+
+* [ADR 57](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-057-RPC.md): 
+Discussion of alternative RPC frameworks for Tendermint.
+* [Issue 1319](https://github.com/tendermint/tendermint/issues/1319): More context 
+on signature aggregation.
+* [Issue 2187](https://github.com/tendermint/tendermint/issues/2187): More context
+on discussions of faster hashing algorithms in Tendermint.
+* [ABCI Commit description](https://github.com/tendermint/spec/blob/master/spec/abci/apps.md#commit):
+More context on locking the mempool during commit.
+* [ABCI local client code](https://github.com/tendermint/tendermint/blob/511bd3eb7f037855a793a27ff4c53c12f085b570/abci/client/local_client.go#L84):
+Clear look at lack of serialization in the local client.
+
+## Discussion
+
+This section attempts to delineate the different sections of Tendermint functionality
+that are often cited as having performance issues. It raises questions and suggests
+lines of inquiry that may be valuable for better understanding Tendermint's performance issues.
+
+As a note: We should avoid quickly adding many microbenchmarks or package level benchmarks. 
+These are prone to being worse than useless as they can obscure what _should_ be  
+focused on: performance of the system from the perspective of a user. We should, 
+instead, tune performance with an eye towards actions user needs.
+
+REQUEST: The list below may be incomplete, if there are additional sections that are often
+cited as creating poor performance, please comment so that they may be included.
+
+### P2P
+
+#### Tendermint cannot scale to large numbers of nodes
+
+A complaint has been reported that Tendermint networks cannot scale to large numbers of nodes.
+The listed number of nodes a user reported as causing issue was in the thousands.
+We don't currently have evidence about what the upper-limit of nodes that Tendermint's
+P2P stack can scale to.
+
+We need to more concretely understand the source of issue and determine what layer
+is causing a problem. It's possible that the P2P layer, in the absence of any reactors
+sending data, is perfectly capable of managing thousands of peer connections. For
+a reasonable networking and application setup, thousands of connections should not present any
+issue for the application.
+
+We need more data to understand the problem directly. We want to drive the popularity
+and adoption of Tendermint and this will mean allowing for chains with more validators.
+We should follow up with users experiencing this issue. We may then want to benchmark
+the P2P layer in a few simple ways:
+1. By starting a network with 1000 nodes and sending no messages to determine memory
+and CPU consumption in the presence of 1000 nodes.
+2. By testing two nodes connected to each other via P2P, and attempting to
+generate as much load between the two nodes as possible. This will help inform
+how much load is incurred from message sends in an otherwise unencumbered P2P stack.
```

Comment:
Making software fall over is fun and possible, but not always super illuminating. Perhaps it's better to phrase this in terms of naming an umber of metrics and thresholds that we should start measuring and monitoring, e.g.:
- transaction rate (throughput)
- achievable block rate
- transaction latency
- message latency in the p2p system (time from creation to socket, and from socket to processing.) 

Maybe there are more measures, but once we have these baselines we can run networks of various sizes to see how network size impacts performance. There are a lot of knobs and I think we might be surprised by the ways that network configuration (max peers per node, etc.) might impact performance. 

This doesn't, of course, work for syncing/catchup, but it's a start, and tracking optimal performance over time, is probably more useful than tracking the "fall over point"

Reply:
Strong agree: Rates and latencies are fantastically illuminating for distributed system performance. (Load testing isn't bad, but I wouldn't start there)

Reply:
I think this is a good direction. Would you mind clarifying what you mean by "achievable block rate"? 

> There are a lot of knobs and I think we might be surprised by the ways that network configuration (max peers per node, etc.) might impact performance.

Definitely agree, I'm really excited to see how different configurations and interventions impact performance.